### PR TITLE
Fix issues in the exam mode test runs

### PR DESCRIPTION
**Pull request by anditurdiu**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
We had a recurring issue where the user object sent from the client was stored on the database on a cascade operation. 

### Description
1. Remove the `cascade.Merge` on `Result.assessor`. 
2. Filter out the results from the submissions of student exams during conduction
3. Add a test_run flag for participations, adapt the `isTestRunParticipation` helper method. (In follow up PRs we will only use this flag to identify test run participations and their submissions)
3. Mark the test run submissions AFTER the instructor submits the test run & evaluate the quiz submission, instead of before.
4. Warn the instructor to delete the test runs before the conduction.
5. In case the instructor forgets to delete the testruns, delete all test runs automatically once the instructor assesses all unsubmitted exams.
6. Do not show second correction table for test runs (only the owner of the test runs can assess the submission once & override it)

### Steps for Testing
1. Create an exam with all exercise types
2. Participate in test runs  and submit. **IMPORTANT**: wait for the build result to return from bamboo!
3. Make sure that the submissions don't appear in the assessment dashboard 
4. Assess the test run submissions
5. Add complaints to the test run submissions
6. Review the complaints
7. View the summary page of the test run, make sure that the results appear correctly. 
8. Make sure that the test runs don't appear in the exam scores
9. Start a test run and do not submit
10. Participate in the exam as a student
11. Assess all unsubmitted exams (button in the student exams tab)
12. Verify that all test runs are deleted.

13. Assess the exam submissions
14. Override with an instructor account
15. Complain about the submissions
17. Review complaints 
18. view scores

### Test Coverage
`StudentExamService` 90%
`ExamService` 91%
`ParticipationService` 72% -> After a followup rework of identifying test run participations will increase to > 80%

### Screenshots
image

### CODE REVIEW

**Review by airelawaleria — COMMENTED** [review]
Tested with @marvinchimeka on ts1. 

1. (minor) After the conducted test run, opening the usual exercise dashboard for the exam displays a translation-not-found error
image

2. (major) **After the test run submission, it is still possible to change the submission in the "View Result" page of the Submission Summary. This results in the submission for that exercise not be accessible from the assessment dashboard of the test runs**
image

3. When firstly assessing the test run, in the assessment dashboard it says "Continue assessment" instead of "Start assessment" (may be expected)

Everything else worked fine.

**Review by marvinchimeka — COMMENTED** [review]
Tested in testing session. Confirm what @airelawaleria said

**Review by JohannesWeiss — COMMENTED** [review]
Tested on ts1 as well. I did not encounter the first of the issues mentioned above, but can confirm the second and third one during test runs. The rest seemed to work fine for me.

**Comment by anditurdiu** [comment]
@airelawaleria @marvinchimeka @JohannesWeiss Thank you for testing. 

1. This is unrelated to this PR
2. Good catch, I fixed it!
3. This is expected behavior for now, because we add results to identify test run submissions. In a followup PR we will fix this

**Comment by FrankeLukas** [comment]
Screenshot from the testing session: 
grafik

**Review by nicolasruscher — CHANGES_REQUESTED** [review]
During the testing session, we found the following bug: Within a exam test-run programming exercise submitting  before receiving an automatic result from bamboo, the manual assessment of the test-run programming exercise is not possible, because the result is missing.

**Review by FuchsDominik — CHANGES_REQUESTED** [review]
We found a bug during the testing session. When trying to click "assess unsubmitted student exams" an error 500 is thrown

**Review by FrankeLukas — CHANGES_REQUESTED** [review]
Error message from the issue mentioned above:
grafik

**Comment by anditurdiu** [comment]
> We found a bug during the testing session. When trying to click "assess unsubmitted student exams" an error 500 is thrown

This is now fixed in: f78524c

**Comment by anditurdiu** [comment]
> Screenshot from the testing session:
> grafik

This is now fixed in: 2b04dfb

**Comment by anditurdiu** [comment]
> During the testing session, we found the following bug: Within a exam test-run programming exercise submitting before receiving an automatic result from bamboo, the manual assessment of the test-run programming exercise is not possible, because the result is missing.

We discussed this with @krusche and decided that we will cover this edge case in a follow up, with the refactoring of the test runs.

**Review by airelawaleria — COMMENTED** [review]
When evaluating complaints, there are no button to accept/reject the complaint, the complaint text and the submission itself are missing.

Also see what @JohannesWeiss said below about submission of complaints and bad request.

image

P.S. the previous bug with editing the submission after the test run is resolved

**Review by simonzilker — CHANGES_REQUESTED** [review]
What @airelawaleria said. When assessing the complaint for a modeling exercise, the buttons do not show up. (Show screenshot)
image

**Review by JohannesWeiss — COMMENTED** [review]
During testing on ts1 the complaint feature did not work for a text exercise. Submitting the complaint does not work and when returning to the summary "Bad request" errors are shown.

Screenshot 2021-01-26 at 17 12 24

**Review by marvinchimeka — CHANGES_REQUESTED** [review]
Tested in testing session. Add the same issues like @airelawaleria and @simonzilker. In additional i could not complain like @JohannesWeiss mentioned.

**Comment by anditurdiu** [comment]
Thanks for testing, I could not reproduce this issue. 
Can you give me some more details? Is this only for test runs? 
If it was for the exam, did you active the student review period? (you have to add the start and end date of the review period to unlock the complaint button)

**Comment by anditurdiu** [comment]
@JohannesWeiss do you have to console output of the bad request? Or the name of the call that is causing it?

File: src/main/java/de/tum/in/www1/artemis/web/rest/StudentExamResource.java [hunk 1]
```diff
@@ -273,6 +275,9 @@ public StudentExamResource(ExamAccessService examAccessService, StudentExamServi
 
         prepareStudentExamForConduction(request, currentUser, testRun);
 
+        // TODO: filter results in participations and submissions before the end of the test run
```
Comment (derLalla) [inline src/main/java/de/tum/in/www1/artemis/web/rest/StudentExamResource.java:278]: is this TODO still needed?

**Comment by anditurdiu** [comment]
> When evaluating complaints, there are no button to accept/reject the complaint, the complaint text and the submission itself are missing.
> 
> Also see what @JohannesWeiss said below about submission of complaints and bad request.
> 
> image
> 
> P.S. the previous bug with editing the submission after the test run is resolved

Culd you please list how you accessed this screen? From the screenshot I can see that there is no assessor set, which should not happen for test runs

**Comment by JohannesWeiss** [comment]
> @JohannesWeiss do you have to console output of the bad request? Or the name of the call that is causing it?

The console output was a HTTP 400 error. After changing the student review period and testing again, the call worked. It seems, that this was the problem. The complaint button was active before setting that period, though.

**Review by derLalla — APPROVED** [review]
I have looked through the code and just found a small `todo` issue. Besided that the code quality is really good! 👍

File: src/main/java/de/tum/in/www1/artemis/web/rest/StudentExamResource.java (hunk 1 above)
Comment (anditurdiu) [reply src/main/java/de/tum/in/www1/artemis/web/rest/StudentExamResource.java]: Good catch, no this todo is not needed anymore,  we filter within the `prepareStudentExamForConduction`

**Review by derLalla — APPROVED** [review]
Nice refactored code! Re-approve.

**Review by airelawaleria — APPROVED** [review]
Tested in the testing session.

We found one bug that should be fixed in a follow-up pr. When an exam has 2 correction rounds, assessment of complaints is broken. @anditurdiu is notified :)

**Review by simonzilker — APPROVED** [review]
Tested in testing session with @anditurdiu .  What @airelawaleria said.

**Review by EneaGore — APPROVED** [review]

**Review by marvinchimeka — APPROVED** [review]
Tested in testing session with @anditurdiu . confirming what @airelawaleria said

**Review by krusche — APPROVED** [review]

**Pull request merged**